### PR TITLE
fix: fix ds required fields

### LINKED ISSUE #26126: [Bug]: S3 | Save button is disabled when Service provider dropdown is changed even once
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

https://github.com/appsmithorg/appsmith/assets/91450662/5e044581-2ba4-47c7-8b6a-eaf5980be2ba

### Steps To Reproduce

1. Create new S3 ds
2. Change the S3 service provider dropdown from `Amazon S3` to some other value
3. Now go back to `Amazon S3` & enter required field values & test ds
4. **Observe - DS does not allow user to save the DS**
5. **Expected - the current dropdown selction should be honoured & DS test/save to behave accordingly**

### Public Sample App

_No response_

### Environment

Production

### Issue video log

_No response_

### Version

Prod & Release v1.9.32

### PR DESCRIPTION
**Pull request by Irongade**
When user toggles between different datasource configuration options in S3, the required fields was not being updated which prevents the user from saving the datasource. This also solves this problem for all possible datasources that have different configuration options.

Fixes #26126 
> if no issue exists, please create an issue and ask the maintainers about this first

- Bug fix (non-breaking change which fixes an issue)

>
>
>
- [x] Manual
- [ ] JUnit
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [x] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [x] Test plan has been peer reviewed by project stakeholders and other QA members
- [x] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/cypress/e2e/Regression/ClientSide/BugTests/Bug26126_spec.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import {
+  dataSources,
+  entityExplorer,
+} from "../../../../support/Objects/ObjectsCore";
+
+describe("Fix DS button disability", function () {
+  it("ensures save button is correctly updated when DS required fields change", function () {
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/Regression/ClientSide/BugTests/Bug26126_spec.ts:7]: Add Bug id to title.

**Review by Aishwarya-U-R — COMMENTED** [review]
Approving Cypress changes,

**Review by sneha122 — DISMISSED** [review]

File: app/client/src/pages/Editor/DataSourceEditor/index.tsx [hunk 2]
```diff
@@ -134,6 +141,8 @@ interface ReduxStateProps {
   defaultKeyValueArrayConfig: Array<string>;
   initialValue: Datasource | ApiDatasourceForm | undefined;
   showDebugger: boolean;
+  featureFlags?: FeatureFlags;
+  currentEnvironment: string;
```
Comment (ayushpahwa) [inline app/client/src/pages/Editor/DataSourceEditor/index.tsx:145]: Is the requirement here to fetch the current environment of the app or the current editing environment. If viewmode is true, then it should be current environment, if not, then it should be current editing environment which in this case can be used from state.filterparams.id

File: app/client/src/pages/Editor/DataSourceEditor/index.tsx [hunk 3]
```diff
@@ -281,6 +290,8 @@ class DatasourceEditorRouter extends React.Component<Props, State> {
       this.props.initializeFormWithDefaults(this.props.pluginType);
     }
 
+    // console.log("hereee", prevProps, this.props);
```
Comment (ayushpahwa) [inline app/client/src/pages/Editor/DataSourceEditor/index.tsx:293]: Please remove this

**Review by ayushpahwa — CHANGES_REQUESTED** [review]

**Review by ayushpahwa — APPROVED** [review]

**Pull request merged**